lib/remote/configstageshandler.cpp

```diff
@@ -132,9 +134,25 @@ void ConfigStagesHandler::HandlePost(
 		if (reload && !activate)
 			BOOST_THROW_EXCEPTION(std::invalid_argument("Parameter 'reload' must be false when 'activate' is false."));
 
-		if (m_RunningPackageUpdates.exchange(true)) {
-			return HttpUtility::SendJsonError(response, params, 423,
-				"Conflicting request, there is already an ongoing package update in progress. Please try it again later.");
+		double currentReloadFailedTime = Application::GetLastReloadFailed();
+		double lastReloadFailedTime = l_LastReloadFailedTime.exchange(currentReloadFailedTime);
+
+		/**
+		 * Once the m_RunningPackageUpdates flag is set, it typically remains set until the current worker process is
+		 * terminated, in which case the new worker will have its own m_RunningPackageUpdates flag set to false.
+		 * However, if the reload fails for any reason, the m_RunningPackageUpdates flag will remain set to true
+		 * in the current worker process, which will prevent any further package updates from being processed until
+		 * the next Icinga 2 restart.
+		 *
+		 * So, in order to prevent such a situation, we are additionally tracking the last time a reload failed
+		 * and allow to bypass the m_RunningPackageUpdates flag only if the last reload failed time was changed
+		 * since the previous request.
+		 */
+		if (m_RunningPackageUpdates.exchange(true) && lastReloadFailedTime == currentReloadFailedTime) {
+			return HttpUtility::SendJsonError(
+				response, params, 423,
+				"Conflicting request, there is already an ongoing package update in progress. Please try it again later."
+			);
 		}
```

Comment:
This has quite some complexity in it due to the interleaved checks and updates of the two atomic variables. Wouldn't this be easier to understand if a mutex was used and there were clear "check if we can continue" and "update the state if we continue" phases?